skorch/tests/callbacks/test_training.py

```diff
@@ -314,3 +314,232 @@ def test_wrong_threshold_mode(
 
         expected_msg = "Invalid threshold mode: 'incorrect'"
         assert exc.value.args[0] == expected_msg
+
+
+
+class TestParamMapper:
+
+    @pytest.fixture
+    def initializer(self):
+        from skorch.callbacks import Initializer
+        return Initializer
+
+    @pytest.fixture
+    def freezer(self):
+        from skorch.callbacks import Freezer
+        return Freezer
+
+    @pytest.fixture
+    def unfreezer(self):
+        from skorch.callbacks import Unfreezer
+        return Unfreezer
+
+    @pytest.fixture
+    def param_mapper(self):
+        from skorch.callbacks import ParamMapper
+        return ParamMapper
+
+    @pytest.fixture
+    def net_cls(self):
+        from skorch import NeuralNetClassifier
+        return NeuralNetClassifier
+
+    @pytest.mark.parametrize('at', [0, -1])
+    def test_subzero_at_fails(self, net_cls, classifier_module,
+                              param_mapper, at):
+        cb = param_mapper(patterns='*', at=at)
+        net = net_cls(classifier_module, callbacks=[cb])
+        with pytest.raises(ValueError):
+            net.initialize()
+
+    @pytest.mark.parametrize('mod_init', [False, True])
+    @pytest.mark.parametrize('weight_pattern', [
+        'sequential.*.weight',
+        lambda name: name.startswith('sequential') and name.endswith('.weight'),
+    ])
+    def test_initialization_is_effective(self, net_cls, classifier_module,
+                                         classifier_data, initializer,
+                                         mod_init, weight_pattern):
+        from torch.nn.init import constant_
+        from skorch.utils import to_numpy
+
+        module = classifier_module() if mod_init else classifier_module
+
+        net = net_cls(
+            module,
+            lr=0,
+            max_epochs=1,
+            callbacks=[
+                initializer(weight_pattern, partial(constant_, val=5)),
+                initializer('sequential.3.bias', partial(constant_, val=10)),
+            ])
+
+        net.fit(*classifier_data)
+
+        assert np.allclose(to_numpy(net.module_.sequential[0].weight), 5)
+        assert np.allclose(to_numpy(net.module_.sequential[3].weight), 5)
+        assert np.allclose(to_numpy(net.module_.sequential[3].bias), 10)
+
+    @pytest.mark.parametrize('mod_init, mod_kwargs', [
+        (False, {}),
+        (True, {}),
+        # Supply a module__ parameter so the model is forced
+        # to re-initialize. Even then parameters should be
+        # frozen correctly.
+        (False, {'module__hidden_units': 5}),
+        (True, {'module__hidden_units': 5}),
+    ])
```

Comment:
This `parametrize` could be split into 2 like in the test above, but this is only a minor point.